core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -596,53 +591,59 @@ namespace kagome::consensus::grandpa {
     BOOST_ASSERT(finalizable());
 
     auto block = finalized_.value();
-    const auto &precommit_justification =
-        getPrecommitJustification(block, precommits_->getMessages());
 
-    SL_DEBUG(logger_,
-             "Round #{}: Finalizing on block #{} hash={}",
-             round_number_,
-             block.number,
-             block.hash.toHex());
+    SL_DEBUG(
+        logger_, "Round #{}: Finalizing on block {}", round_number_, block);
 
     GrandpaJustification justification{
         .round_number = round_number_,
         .block_info = block,
-        .items = std::move(precommit_justification)};
+        .items = getPrecommitJustification(block, precommits_->getMessages())};
 
-    if (need_to_notice_at_finalizing_) {
-      sendFinalize(block, std::move(justification));
-    } else {
-      auto res = env_->finalize(voter_set_->id(), justification);
+    auto res = env_->finalize(voter_set_->id(), justification);
+    if (res.has_error()) {
+      SL_WARN(logger_,
+              "Round #{}: Finalizing on block {} is failed",
+              round_number_,
+              block);
+    }
+
+    // Note: this is hack for behavior like in substrate
```

Comment:
Good to leave a link to github at the specific revision of substrate with the corresponding source

Reply:
Substrate has very different design. This way was found by the watching (logs)